### PR TITLE
Reuse write request from distributor to Ingesters

### PR DESCRIPTION
**Pull request by alanprot**
**What this PR does**:

Just some tentative to reuse the write requests and the byte slices when marshaling the write requests from distributors to ingesters.

Bench Results: 

```
BenchmarkMarshallWriteRequest/no-pool-16         	   40658	     29358 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/no-pool-16         	   40474	     29386 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/no-pool-16         	   40921	     29294 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/no-pool-16         	   40840	     29317 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/no-pool-16         	   40756	     29344 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/byte_pool-16       	   44792	     26797 ns/op	     104 B/op	       1 allocs/op
BenchmarkMarshallWriteRequest/byte_pool-16       	   44408	     26797 ns/op	     106 B/op	       1 allocs/op
BenchmarkMarshallWriteRequest/byte_pool-16       	   44719	     26799 ns/op	     105 B/op	       1 allocs/op
BenchmarkMarshallWriteRequest/byte_pool-16       	   44761	     26792 ns/op	     105 B/op	       1 allocs/op
BenchmarkMarshallWriteRequest/byte_pool-16       	   44568	     26796 ns/op	     105 B/op	       1 allocs/op
BenchmarkMarshallWriteRequest/byte_and_write_pool-16         	   44536	     26646 ns/op	       0 B/op	       0 allocs/op
BenchmarkMarshallWriteRequest/byte_and_write_pool-16         	   45062	     26622 ns/op	       0 B/op	       0 allocs/op
BenchmarkMarshallWriteRequest/byte_and_write_pool-16         	   44780	     26826 ns/op	       0 B/op	       0 allocs/op
BenchmarkMarshallWriteRequest/byte_and_write_pool-16         	   45090	     26625 ns/op	       0 B/op	       0 allocs/op
BenchmarkMarshallWriteRequest/byte_and_write_pool-16         	   44688	     26822 ns/op	       0 B/op	       0 allocs/op
```

```
## no-pool x byte_pool
name                     old time/op    new time/op    delta
MarshallWriteRequest-16    29.3µs ± 0%    26.8µs ± 0%   -8.67%  (p=0.008 n=5+5)

name                     old alloc/op   new alloc/op   delta
MarshallWriteRequest-16    19.1kB ± 0%     0.1kB ± 1%  -99.45%  (p=0.008 n=5+5)

name                     old allocs/op  new allocs/op  delta
MarshallWriteRequest-16      2.00 ± 0%      1.00 ± 0%  -50.00%  (p=0.008 n=5+5)



## byte_pool x byte_and_write_pool
name                     old time/op    new time/op    delta
MarshallWriteRequest-16    26.8µs ± 0%    26.7µs ± 0%      ~     (p=0.651 n=5+5)

name                     old alloc/op   new alloc/op   delta
MarshallWriteRequest-16      105B ± 1%        0B       -100.00%  (p=0.008 n=5+5)

name                     old allocs/op  new allocs/op  delta
MarshallWriteRequest-16      1.00 ± 0%      0.00       -100.00%  (p=0.008 n=5+5)


## no-pool x byte_and_write_pool
name                     old time/op    new time/op    delta
MarshallWriteRequest-16    29.3µs ± 0%    26.7µs ± 0%    -8.97%  (p=0.008 n=5+5)

name                     old alloc/op   new alloc/op   delta
MarshallWriteRequest-16    19.1kB ± 0%     0.0kB       -100.00%  (p=0.008 n=5+5)

name                     old allocs/op  new allocs/op  delta
MarshallWriteRequest-16      2.00 ± 0%      0.00       -100.00%  (p=0.008 n=5+5)
```

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [X] Tests updated
- [ ] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by yeya24 — COMMENTED** [review]
Can you add a Benchmark test for this?

**Comment by alanprot** [comment]
> Can you add a Benchmark test for this?

Yup! will do

**Comment by alanprot** [comment]
Here is the branchmark: We are basically avoiding reallocating the byte[] when serializing.

```
BenchmarkMarshallWriteRequest/no-pool-12         	   51582	     23269 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/no-pool-12         	   53248	     22926 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/no-pool-12         	   52886	     22982 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/no-pool-12         	   51668	     23563 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/no-pool-12         	   51082	     22742 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/pool-12            	   57105	     20625 ns/op	      24 B/op	       1 allocs/op
BenchmarkMarshallWriteRequest/pool-12            	   58096	     20151 ns/op	      24 B/op	       1 allocs/op
BenchmarkMarshallWriteRequest/pool-12            	   59668	     20219 ns/op	      24 B/op	       1 allocs/op
BenchmarkMarshallWriteRequest/pool-12            	   59752	     20385 ns/op	      24 B/op	       1 allocs/op
BenchmarkMarshallWriteRequest/pool-12            	   59164	     20557 ns/op	      24 B/op	       1 allocs/op
```

**Comment by alanprot** [comment]
I could make no have any allocation at all:

```
BenchmarkMarshallWriteRequest/no-pool-12         	   50653	     23301 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/no-pool-12         	   46936	     24360 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/no-pool-12         	   52663	     24185 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/no-pool-12         	   52234	     24141 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/no-pool-12         	   51400	     23053 ns/op	   19136 B/op	       2 allocs/op
BenchmarkMarshallWriteRequest/pool-12            	   58034	     21088 ns/op	       0 B/op	       0 allocs/op
BenchmarkMarshallWriteRequest/pool-12            	   58436	     20775 ns/op	       0 B/op	       0 allocs/op
BenchmarkMarshallWriteRequest/pool-12            	   58020	     20741 ns/op	       0 B/op	       0 allocs/op
BenchmarkMarshallWriteRequest/pool-12            	   57319	     20861 ns/op	       0 B/op	       0 allocs/op
BenchmarkMarshallWriteRequest/pool-12            	   54622	     23412 ns/op	       0 B/op	       0 allocs/op
```

**Review by yeya24 — APPROVED** [review]
I am up for the change. Very nice improvements!

File: pkg/cortexpb/slicesPool.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+		index = 0
+	}
+
+	s := sp.pools[index].Get().(*[]byte)
+	*s = (*s)[:size]
+	return s
+}
+
+func (sp *byteSlicePools) reuseSlice(s *[]byte) {
+	index := int(math.Floor(math.Log2(float64(cap(*s))))) - minPoolSizePower
```
Comment (alvinlin123) [inline pkg/cortexpb/slicesPool.go:53]: Should we move this to a helper function?

File: pkg/cortexpb/slicesPool.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+	if index >= len(sp.pools) {
+		return
+	}
+
+	if index < 0 {
+		index = 0
```
Comment (alvinlin123) [inline pkg/cortexpb/slicesPool.go:60]: If a caller did not follow the contract and give you a slice with capacity less than 2^5, it kinda of break the invariant that the pool should have a min slice of size 2^5 right? 

Would this be a problem?

File: pkg/cortexpb/timeseries.go [hunk 3]
```diff
@@ -72,6 +82,30 @@ func (p *PreallocTimeseries) Unmarshal(dAtA []byte) error {
...
+	return dAtA[:n], nil
+}
+
+func ReuseWriteRequest(req *PreallocWriteRequest) {
+	if req.data != nil {
+		bytePool.reuseSlice(req.data)
+		req.data = nil
+	}
+
+	writeRequestPool.Put(req)
```
Comment (alvinlin123) [inline pkg/cortexpb/timeseries.go:102]: For safety, should we clear out all fields of `writeRequestPool` before putting it back to the pool?

File: pkg/cortexpb/slicesPool.go (hunk 2 above)
Comment (alanprot) [reply pkg/cortexpb/slicesPool.go:60]: Good catch! thanks

File: pkg/cortexpb/slicesPool.go (hunk 1 above)
Comment (alanprot) [reply pkg/cortexpb/slicesPool.go:53]: One is Ceil and the other is floor.

On getSlice i wanna get a slice with capacity > than the one i'm allocating (hence ceil).

On the reuseSlice i wanna put the slice on the pool that is AT LEAST the capacity of the slice that i'm reusing:


Ex: getSlice(250) => log2 250 = 7.96 => 2^8 = get from pool with cap = 256

ReuseSlice(250) => log2 250 = 7.96 -> 2^7 => return to the pool with cap(128)

In the case of 256 for instance the pool used will be 256 on both cases.

File: pkg/cortexpb/timeseries.go (hunk 3 above)
Comment (alanprot) [reply pkg/cortexpb/timeseries.go:102]: +1

File: pkg/cortexpb/slicesPool.go (hunk 1 above)
Comment (alvinlin123) [reply pkg/cortexpb/slicesPool.go:54]: oh ok I see

**Review by alvinlin123 — APPROVED** [review]

**Pull request merged**